File: plugin/src/main/java/org/godotengine/plugin/vr/oculus/mobile/api/OvrPerformance.kt
/**
 * Wrapper for the VrApi performance apis.
 */
@file:JvmName("OvrPerformance")

package org.godotengine.plugin.vr.oculus.mobile.api

import org.godotengine.plugin.vr.oculus.mobile.OvrMobilePlugin

/**
 * Extra latency mode pipelines app CPU work a frame ahead of VR composition.
 */
enum class ExtraLatencyMode(internal val value: Int) {
    OFF(0),
    ON(1),
    DYNAMIC(2);
}

/**
 * Used to control swapchain foveation levels.
 */
enum class FoveationLevel(internal val value: Int) {
    NONE(0),
    LOW(1),
    MEDIUM(2),
    HIGH(3),
    HIGH_TOP(4);
}

/**
 * Set the CPU and GPU performance levels.
 * Increasing the levels increases performance at the cost of higher power consumption
 * which likely leads to a greater chance of overheating.
 *
 * Levels will be clamped to the expected range.
 * Default clock levels are cpuLevel = 2, gpuLevel = 2.
 */
fun OvrMobilePlugin.setClockLevels(cpuLevel: Int, gpuLevel: Int): Boolean {
    return if (isSharedLibLoaded()) {
        nativeSetClockLevels(cpuLevel, gpuLevel)
    } else {
        false
    }
}

/**
 * If [ExtraLatencyMode.ON] specified, adds an extra frame of latency for full GPU
 * utilization. Default is [ExtraLatencyMode.OFF].
 */
fun OvrMobilePlugin.setExtraLatencyMode(latencyMode: ExtraLatencyMode): Boolean {
    return if (isSharedLibLoaded()) {
        nativeSetExtraLatencyMode(latencyMode.value)
    } else {
        false
    }
}

/**
 * Set the fixed foveated rendering level if it's available.
 * Returns true if it's available and the given level is valid, false otherwise.
 *
 * **Note:** Dynamic foveation should be used instead.
 *
 * [Additional docs](https://developer.oculus.com/documentation/quest/latest/concepts/mobile-ffr/)
 */
fun OvrMobilePlugin.setFoveationLevel(foveationLevel: FoveationLevel): Boolean {
    return if (isSharedLibLoaded()) {
        nativeSetFoveationLevel(foveationLevel.value)
    } else {
        false
    }
}

/**
 * Sets the swap interval to control the application frame timing.
 *
 * [Additional docs](https://developer.oculus.com/documentation/mobilesdk/latest/concepts/mobile-vrapi#frame-timing)
 */
fun OvrMobilePlugin.setSwapInterval(swapInterval: Int): Boolean {
    return if (isSharedLibLoaded()) {
        nativeSetSwapInterval(swapInterval)
    } else {
        false
    }
}

/**
 * Enable dynamic foveation.
 *
 * [Additional docs](https://developer.oculus.com/documentation/native/android/mobile-ffr/)
 */
fun OvrMobilePlugin.enableDynamicFoveation(enable: Boolean): Boolean {
    return if (isSharedLibLoaded()) {
        nativeEnableDynamicFoveation(enable)
    } else {
        false
    }
}

private external fun nativeEnableDynamicFoveation(enable: Boolean): Boolean

private external fun nativeSetClockLevels(cpuLevel: Int, gpuLevel: Int): Boolean

private external fun nativeSetExtraLatencyMode(latencyMode: Int): Boolean

private external fun nativeSetFoveationLevel(foveationLevel: Int): Boolean

private external fun nativeSetSwapInterval(swapInterval: Int): Boolean

File: plugin/src/main/cpp/api/ovr_performance.cpp
#include "ovr_performance.h"
#include "api_common.h"
#include <vector>

namespace ovrmobile {

OvrPerformance::OvrPerformance() = default;

OvrPerformance::~OvrPerformance() = default;

void OvrPerformance::_init() {}

void OvrPerformance::_register_methods() {
    register_method("set_clock_levels", &OvrPerformance::set_clock_levels);
    register_method("set_extra_latency_mode", &OvrPerformance::set_extra_latency_mode);
    register_method("set_foveation_level", &OvrPerformance::set_foveation_level);
    register_method("set_swap_interval", &OvrPerformance::set_swap_interval);
    register_method("set_enable_dynamic_foveation", &OvrPerformance::set_enable_dynamic_foveation);
}

bool set_clock_levels(OvrMobileSession *session, int cpu_level, int gpu_level) {
    return check_session_initialized<bool>(session, [&]() {
        return vrapi_SetClockLevels(session->get_ovr_mobile_context(), cpu_level, gpu_level) ==
               ovrSuccess;
    });
}

bool set_extra_latency_mode(OvrMobileSession *session, ovrExtraLatencyMode latency_mode) {
    return check_session_initialized<bool>(session, [&]() {
        return vrapi_SetExtraLatencyMode(session->get_ovr_mobile_context(), latency_mode) ==
               ovrSuccess;
    });
}

bool set_foveation_level(OvrMobileSession *session, int foveation_level) {
    return check_session_initialized<bool>(session, [&]() {
        const ovrJava *ovr_java = session->get_ovr_java();
        // Check if foveation is available.
        int foveation_available =
                vrapi_GetSystemPropertyInt(ovr_java, VRAPI_SYS_PROP_FOVEATION_AVAILABLE);
        if (foveation_available == VRAPI_TRUE) {
            // Foveation level for the Oculus Quest ranges from 0 to 4
            // See https://developer.oculus.com/documentation/quest/latest/concepts/mobile-ffr/
            bool foveation_valid = foveation_level >= 0 && foveation_level <= 4;

            if (foveation_valid) {
                vrapi_SetPropertyInt(ovr_java, VRAPI_FOVEATION_LEVEL, foveation_level);
            }
            return foveation_valid;
        }
        return false;
    });
}

bool set_swap_interval(OvrMobileSession *session, unsigned int swap_interval) {
    return check_session_initialized<bool>(session, [&]() {
        if (swap_interval == 1 || swap_interval == 2) {
            session->set_swap_interval(swap_interval);
            return true;
        }
        return false;
    });
}

bool set_enable_dynamic_foveation(OvrMobileSession *session, bool enable) {
    return check_session_initialized<bool>(session, [&]() {
        const ovrJava *ovr_java = session->get_ovr_java();
        // Check if foveation is available.
        int foveation_available =
                vrapi_GetSystemPropertyInt(ovr_java, VRAPI_SYS_PROP_FOVEATION_AVAILABLE);
        if (foveation_available == VRAPI_TRUE) {
            vrapi_SetPropertyInt(ovr_java, VRAPI_DYNAMIC_FOVEATION_ENABLED, enable);
            return true;
        }
        return false;
    });
}

} // namespace ovrmobile
